### ISSUE TITLE
AsyncFtpClient is disposed off before other methods can execute

### ISSUE THREAD

**Issue by GW-Lab** [robinrodricks/FluentFTP#1634]
**FTP Server OS:** Unix / Windows / Embedded

**FTP Server Type:** Vsftpd / glFTPd / BFTPd / ProFTPD / Pure-FTPd / IBM CS FTP (z/OS, OS/400) / Windows CE / Windows Server IIS / Vax / VMS / OpenVMS / Tandem / HP NonStop

**Client Computer OS:** Windows / Ubuntu / Debian / Linux Mint / Arch Linux

**FluentFTP Version:** ?

**Framework:** .NET 2 / 3.5 / 4 / 5 / 6 / 7 / 8 / UWP / Xamarin / Mono

<write details about your bug report / feature request here>

**Logs :**

```

<paste VERBOSE logs here>

```

**Comment by GW-Lab** [comment]
Used:
C# 12 .Net 8.07
Unix FTP server.

**Comment by GW-Lab** [comment]
Hallo Robin,

Thanks for your reaction.

Herewith a minimal sample to illustrate/reproduce the problem.

I am a little bit puzzled about this hopefully there is a simple explanation what I am doing wrong or don’t understand.

Tools used

  *   Windows 11 pro build 22631.4037
  *   Visual Studio 2022 version: 17.11.1
  *   C# 12
  *   .Net 8.07 -> Winforms application <-
  *   FluentFTP ver 49.0.1 and 51.0.0
  *   Host FTP server linux

--------------------------------------------------- Minimal source code  sample ----------------------------------------------------------------

The source code that is not returning to the caller (FluentFTP ver: 51.0.0) but is running fine with FluentFTP ver: 49.0.1

// ==============================  Caller ================================

private async void Form1_Load(object sender, EventArgs e)
{
      var ftpClient = new FTPClientAsync(host, userName, password);

      if (await ftpClient.LoggedInAsync())
      {
            var a = 1; // This line is executed in the debugger(FluentFTP ver: 49.0.1) This line is never executed (FluentFTP ver: 51.0.0)
      }
      else
      {
            var b = 1; // This line is never executed in the debugger (FluentFTP ver: 49.0.1) because login credentials are OK

      }
}

// ==============================  CLASS ================================

using FluentFTP;   // Implement FluentFTP (reason: starting with ".Net version 9.0" support for FtpWebRequest will be dropped

namespace FTPTest;

public class FTPClientAsync(string hostIP, string userName, string password)
{
      private readonly string hostIP = hostIP;
      private readonly string userName = userName;
      private readonly string passWord = password;

      public async Task<bool> LoggedInAsync()
      {
            var token = new CancellationToken();
            using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.passWord);
            await ftp.Connect(token);

            if (ftp.IsConnected)
            {
                  return true;   // This line is executed in de debugger but no return to the caller (FluentFTP ver: 49.0.1 and FluentFTP ver: 51.0.0)

            }
            else
            {
                  return false;
            };
      }
}

Kind Regards

Gijs Wassink

**Comment by GW-Lab** [comment]
Hallo Robin,

Thanks for your reaction.

Sorry I made a mistake in the previous e-mail please discard.

Herewith a minimal sample to illustrate/reproduce the problem.

I am a little bit puzzled about this hopefully there is a simple explanation what I am doing wrong or don’t understand.

Tools used

  *   Windows 11 pro build 22631.4037
  *   Visual Studio 2022 version: 17.11.1
  *   C# 12
  *   .Net 8.07 -> Winforms application <-
  *   FluentFTP ver 49.0.1 and 51.0.0
  *   Host FTP server linux

--------------------------------------------------- Minimal source code  sample ----------------------------------------------------------------

The source code that is not returning to the caller (FluentFTP ver: 51.0.0) but is running fine with FluentFTP ver: 49.0.1

// ==============================  Caller ================================

private async void Form1_Load(object sender, EventArgs e)
{
      var ftpClient = new FTPClientAsync(host, userName, password);

      if (await ftpClient.LoggedInAsync())
      {
            var a = 1; // This line is executed in the debugger(FluentFTP ver: 49.0.1) This line is never executed (FluentFTP ver: 51.0.0)
      }
      else
      {
            var b = 1; // This line is never executed in the debugger (FluentFTP ver: 49.0.1) because login credentials are OK

      }
}

// ==============================  CLASS ================================

using FluentFTP;   // Implement FluentFTP (reason: starting with ".Net version 9.0" support for FtpWebRequest will be dropped

namespace FTPTest;

public class FTPClientAsync(string hostIP, string userName, string password)
{
      private readonly string hostIP = hostIP;
      private readonly string userName = userName;
      private readonly string passWord = password;

      public async Task<bool> LoggedInAsync()
      {
            var token = new CancellationToken();
            using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.passWord);
            await ftp.Connect(token)<ftp://ftp.Connect(token)>;

            if (ftp.IsConnected<ftp://ftp.IsConnected>)
            {
                  return true;   // This line is executed in de debugger but no return to the caller (FluentFTP ver: 51.0.0)

            }
            else
            {
                  return false;
            };
      }
}

Kind Regards

Gijs Wassink

**Comment by FanDjango** [comment]
It just hangs? How long did you wait? Can you post a verbose log?

**Comment by GW-Lab** [comment]
With version: 49.01 it return within a second with 51.0.0 I wait more than 30 seconds.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by FanDjango** [comment]
Ok, I would like a log, verbose, from the PWD command to the end. I want to see the dispose/close async logic...

**Comment by FanDjango** [comment]
Note that your

` using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.passWord);`

client gets disposed as part of the return. Which means it gets closed, connection disconnected.

**Comment by FanDjango** [comment]
If it worked before, it was because much of the async disconnect/close/dispose logic ran as sync and not async.

We will find out more tomorrow.

**Comment by GW-Lab** [comment]
public async Task<bool> LoggedInAsync()
	{
		var token = new CancellationToken();
------>	**using** var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.passWord);
		await ftp.Connect(token);

		if (ftp.IsConnected)
		{
			return true;   // This line is executed but no return to the caller
		}
		else
		{
			return false;
		};
	}

Correct: If I remove the ------> **using** then it returns normal (no hanging)

**Comment by FanDjango** [comment]
Well, I think what is happening, is that the real intensive cleanup of sync/async close/disconnect/dispose is showing up "unclean" code (in FluentFTP, "mis"used in users code) - so you are unfortunately a victim.

If you now look at the log, you will see that after the return, there is no more client, no more connection, no nothing for you to continue working with (if that is your intent).

**Comment by FanDjango** [comment]
Also, with .NET 8, you would use:

```cs
await using(...............) {


}
```

**Comment by GW-Lab** [comment]
public async Task<bool> LoggedInAsync()
{
	var token = new CancellationToken();
	await using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.passWord);
	await ftp.Connect(token);

	if (ftp.IsConnected)
	{
		return true;   // This line is executed but no return to the caller
	}
	else
	{
		return false;
	};
}

I think this is also working-:)
With this knowledge I will now implement the other methods I have.

Thank you very much for the support is a great help.

**Comment by FanDjango** [comment]
Look, I cannot reproduce it. I tried this:

```cs
		static void My_SmallUploadBytes_Test() {
			bool b = My_SmallUploadBytes_Test1Async().GetAwaiter().GetResult();   <---- I get bool b returned no problem
		}

		static async Task<bool> My_SmallUploadBytes_Test1Async() {
			using var client = new AsyncFtpClient("ftps", "bla", "blabla");
			client.Config.EncryptionMode = FtpEncryptionMode.Explicit;
			client.Config.DataConnectionType = FtpDataConnectionType.PASV;
			client.Config.DownloadDataType = FtpDataType.Binary;
			client.Config.ValidateAnyCertificate = true;
			client.Config.Noop = true;
			client.Config.NoopInterval = 750;
			client.Config.NoopTestConnectivity = false;

			client.Config.Navigate = FtpNavigate.SemiAuto;

			await client.Connect();

			if (client.IsConnected) {
				Console.WriteLine("HERE THE RETURN STATEMENT IS executed");
				return true;
			}
			else {
				return false;
			}
		}
```
Even **without the await** for the using.

```
>         Connect(False)
Status:   FluentFTP 51.0.0.0(.NET 5.0) AsyncFtpClient
Status:   Connecting(async) AsyncFtpClient.FtpSocketStream(control) IP #1 = ***:21
Status:   Waiting for a response
Response: 220 ProFTPD Server (Debian) [::ffff:192.168.1.109] [739118,57d]
Status:   Detected FTP server: ProFTPD
Command:  AUTH TLS
Status:   Waiting for response to: AUTH TLS
Response: 234 AUTH TLS successful [2ms]
Status:   FTPS authentication successful, lib = .NET SslStream, cipher suite = Tls12 (Aes256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, 44550, 255) [95ms]
Command:  USER ***
Status:   Waiting for response to: USER ***
Response: 331 Password required for *** [2ms]
Command:  PASS ***
Status:   Waiting for response to: PASS ***
Response: 230 User *** logged in [42ms]
Command:  PBSZ 0
Status:   Waiting for response to: PBSZ 0
Response: 200 PBSZ 0 successful [<1ms]
Command:  PROT P
Status:   Waiting for response to: PROT P
Response: 200 Protection set to Private [1ms]
Command:  FEAT
Status:   Waiting for response to: FEAT
Response: 211-Features:
Response: AUTH TLS
Response: CCC
Response: CLNT
Response: CSID
Response: EPRT
Response: EPSV
Response: HASH CRC32;MD5;SHA-1*;SHA-256;SHA-512;
Response: HOST
Response: LANG en-US.UTF-8*;en-US
Response: MD5
Response: MDTM
Response: MFF modify;UNIX.group;UNIX.mode;
Response: MFMT
Response: MLST modify*;perm*;size*;type*;unique*;UNIX.group*;UNIX.groupname*;UNIX.mode*;UNIX.owner*;UNIX.ownername*;
Response: PBSZ
Response: PROT
Response: RANG STREAM
Response: REST STREAM
Response: SITE COPY
Response: SITE MKDIR
Response: SITE RMDIR
Response: SITE SYMLINK
Response: SITE UTIME
Response: SIZE
Response: SSCN
Response: TVFS
Response: UTF8
Response: XCRC
Response: XMD5
Response: XSHA
Response: XSHA1
Response: XSHA256
Response: XSHA512
Response: 211 End [5ms]
Status:   Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed
Command:  OPTS UTF8 ON
Status:   Waiting for response to: OPTS UTF8 ON
Response: 200 UTF8 set to on [<1ms]
Command:  SYST
Status:   Waiting for response to: SYST
Response: 215 UNIX Type: L8 [<1ms]
Status:   Active ServerHandler is: ProFTPD
Status:   Listing parser set to: Machine
Command:  PWD
Status:   Waiting for response to: PWD
Response: 257 "/home/mike" is the current directory [<1ms]
Status:   NoopDaemon enabled
>         Dispose()
Status:   NoopDaemon(AsyncFtpClient) is initialized, NoopInterval = 1000ms
HERE THE RETURN STATEMENT IS executed
Status:   Warning: sync dispose called for AsyncFtpClient object
>         DisposeAsync()
Status:   Disposing(async) AsyncFtpClient
>         Disconnect()
Command:  QUIT
Status:   Waiting for response to: QUIT
Response: 221 Goodbye. [1ms]
Status:   NoopDaemon disabled
Status:   Disposing(async) AsyncFtpClient.FtpSocketStream(control)
Status:   Disposing(async) AsyncFtpClient.FtpSocketStream(control) (redundant)
Status:   Waiting for Daemon termination(AsyncFtpClient)
Status:   NoopDaemon terminated
Status:   Daemon terminated
```
See this message: `Status:   Warning: sync dispose called for AsyncFtpClient object`? This will **not** happen if you add the **await** in front of the using. But it still returns fine even without the await.

If you want to debug this further, since you are using an intermediate class, you **NEED** to console.writeln(...) some messages **and** also have the FluentFTP log going to console also. You want to look at the sequence of events and especially the "disposing" stuff at the end.

**Comment by GW-Lab** [comment]
I use a Winforms app not a console app. Is that a crucial difference?

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by GW-Lab** [comment]
Try this create a new Winform app target .Net 8.07

Make the Form_Load event like this. Run the app in debug mode and it will **HANG**

private  async void Form1_Load(object sender, EventArgs e)
{
   using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.password);
  await ftp.Connect();
}

This will execute normal no HANGING
private  async void Form1_Load(object sender, EventArgs e)
{
  await using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.password);
  await ftp.Connect();
}

**Comment by robinrodricks** [comment]
I think what's happening in your HANGING example above, is that the AsyncFtpClient is getting disposed before any subsequent operations can occur. So you will need to use `await using` or else keep a ref to the AsyncFtpClient so it does not get disposed off so quickly.

**Comment by GW-Lab** [comment]
I am fine to use await in front of the using it solves my problem 😉

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by FanDjango** [comment]
@GW-Lab Good one, I'll try that and investigate, especially on .NET versions that do not yet support "await using".

**Comment by FanDjango** [comment]
@robinrodricks And it used to work pre V51, because async dispose was effectively sync, which other users complained about.

**Comment by FanDjango** [comment]
So, **there was a problem** in the dispose code.

@robinrodricks There was a problem in the dispose code. Your explanation end my explanations were wrong. The hang has to do with GetAwaiter().GetResult() on the GUI thread.

The source:

```cs
private async void Form1_Load(object sender, EventArgs e)
{
using var ftp = new AsyncFtpClient(this.hostIP, this.userName, this.password);
await ftp.Connect();
}
```

in a WinForms App will hang exactly on the SYNC dispose of the ASYNC client:

`DisposeAsync()..ConfigureAwait(false).GetAwaiter().GetResult();` is a NOGO.

**Comment by FanDjango** [comment]
By using "await" this bad code is effectively bypassed, nice workaround. **BUT: This is not available to .NET Framework 4.6.2 and 4.7.2. So... needed fixing.

**Comment by robinrodricks** [comment]
Good to know.

**Comment by GW-Lab** [comment]
Thanks for fixing this issue.
When will the fix public available?

**Comment by FanDjango** [comment]
@GW-Lab Currently difficult to say. A few things are being worked on at this moment.

**Comment by FanDjango** [comment]
@GW-Lab Fix released as Nuget V51.1.0, FYI

**Comment by GW-Lab** [comment]
Thank you for fixing

Gijs

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________
